src/software/ai/navigator/path_planner/bang_bang_trajectory_1d.cpp

```diff
@@ -0,0 +1,310 @@
+                       .velocity     = max_vel,
+                       .acceleration = -max_decel});
+}
+
+void BangBangTrajectory1D::generateTriangularTrajectory(
+    double initial_pos, double final_pos, double initial_vel, double max_accel,
+    double max_decel, double time_offset_sec)
+{
+    double direction = std::copysign(1.0, final_pos - initial_pos);
+    double dist      = std::abs(final_pos - initial_pos);
+
+    // Given the following constraints:
+    // - Travel exactly a distance of dist
+    // - Have a final velocity of 0
+    // - Take the least amount of time possible (i.e. accelerate until we have to
+    // decelerate) Calculate the time it takes for us to decelerate from the highest
+    // velocity we can reach to a full stop. The full derivation of the formula can be
+    // found here: https://www.desmos.com/calculator/qvrvtplgk7 Note that the full
+    // derivation also supports a non-zero final velocity, but we currently don't support
+    // that here.
+    double t_decel = std::sqrt((initial_vel * initial_vel + 2 * dist * max_accel) /
+                               (max_decel * (max_accel + max_decel)));
+
+    double signed_accel = std::copysign(max_accel, direction);
+    double signed_decel = -std::copysign(max_decel, direction);
+
+    // Calculate the max velocity we will reach
+    // vf = vi + at  =>  vi = -at  (vf = 0)
+    double v_max_reached = -signed_decel * t_decel;
+
+    // Calculate the time to accelerate to the max reached velocity
+    double t_accel = (v_max_reached - initial_vel) / signed_accel;
+    double d_accel = initial_vel * t_accel + 0.5 * signed_accel * t_accel * t_accel;
+
+    // Add the trajectory parts
+    addTrajectoryPart({.end_time_sec = time_offset_sec + t_accel,
+                       .position     = initial_pos,
+                       .velocity     = initial_vel,
+                       .acceleration = signed_accel});
+    addTrajectoryPart({.end_time_sec = time_offset_sec + t_accel + t_decel,
+                       .position     = initial_pos + d_accel,
+                       .velocity     = v_max_reached,
+                       .acceleration = signed_decel});
+}
+
+double BangBangTrajectory1D::getPosition(double t_sec) const
+{
+    TrajectoryPart traj_part;
+    double t_delta_sec;
+    getTrajPartAndDeltaTime(t_sec, traj_part, t_delta_sec);
+
+    // d = vi * t + 0.5 * a * t^2
+    // p = pi + d
+    return traj_part.position + traj_part.velocity * t_delta_sec +
+           0.5 * traj_part.acceleration * t_delta_sec * t_delta_sec;
+}
+
+double BangBangTrajectory1D::getVelocity(double t_sec) const
+{
+    TrajectoryPart traj_part;
+    double t_delta_sec;
+    getTrajPartAndDeltaTime(t_sec, traj_part, t_delta_sec);
+
+    // vf = vi + at
+    return traj_part.velocity + traj_part.acceleration * t_delta_sec;
+}
+
+double BangBangTrajectory1D::getAcceleration(double t_sec) const
+{
+    t_sec                   = std::clamp(t_sec, 0.0, getTotalTime());
```

Comment:
move the clamp into `getTrajectoryIndexAtTime`

Reply:
Decided to actually remove it, since given the way `getTrajectoryIndexAtTime` is implemented, it already safely handles negative or larger than total time `t_sec` time

Reply:
Actually ended up reverting it to avoid duplicated clamps in `getTrajPartAndDeltaTime`, since `getTrajPartAndDeltaTime` uses the clamped time for some additional calculations as well.